[영화 극장 선택] 호두 1, 2단계 미션 제출합니다.

Junyoung-WON opened the pull request:
# 영화 극장 선택 - Step1, 2

안녕하세요, 두루!
이번에 두루에게 두번째 미션에 대한 피드백을 받게 된 호두입니다~
처음 뵙게 되었네요! 잘 부탁드리겠습니다!

아래는 2번째 미션에 대한 커밋 변경 사항들입니다.

https://github.com/Junyoung-WON/android-movie-theater/compare/e420352348362110056382f945cb74f44e051da0..d8638cf60b737bc23f5f904b7cbb25fd839d52ea

---

## 아직 구현하지 못한 기능

### 극장별 날짜 데이터 적용 X
극장별 날짜와 시간 정보를 만들었지만, 스피너에 해당 데이터를 연결짓지 않았습니다.
추후에 해당 기능을 개선하겠습니다!

### UI 테스트 코드 작성 X
기능 구현이 먼저라 생각해서, UI 테스트는 작성하지 못했습니다.
UI 테스트도 PR 요청 후 작성해나가겠습니다!

kkosang committed:
chore: step 0 시작

kkosang committed:
docs: 기능 요구 사항 작성

kkosang committed:
refactor(repository): Repository interface 패키지 이동
- model/repository -> presentation/repository

kkosang committed:
rename: 예약 할 영화 정보 클래스 네이밍 수정
- ScreeningMovieInfo -> ReservationMovieInfo

kkosang committed:
feat: 극장별 상영 영화 정보를 가져오는 저장소 구현
- Theater, TheaterRepository

kkosang committed:
feat: 영화 선택 시 상영 중인 극장 정보를 보여주는 BottomSheetDialog 구현

View:
 - bottom sheet dialog xml 과 item xml 작성
 - 필요 색상 colors.xml 추가
 - BottomSheetDialogFragment 구현
 - RecyclerView와 Adapter 및 ViewHolder 구현
 - Contract 작성 및 Presenter 구현

Repository:
 - 극장 상영 정보를 불러오는 레포지토리 수정

kkosang committed:
refactor: presentation 패키지 구성 변경

- 화면 단위로 presentation 분리

kkosang committed:
feat: 극장 선택 시 영화 상세 화면으로 전환

kkosang committed:
refactor: 영화 상세 화면 DataBinding 리팩터링

- gradle.build에 데이터 바인딩 설정 추가
- Presenter와 View를 바인딩하여 findViewById 사용 감소 및 Contract 메서드 감소

kkosang committed:
feat: 극장의 정보를 나타내는 Ui Model 작성 및 적용

kkosang committed:
refactor: 극장 선택에 관한 DataBinding 진행

- TheaterUiModel과 onClick 리스너에 대한 바인딩 태그 등록
- ViewHolder에 binding 생성 후 극장 정보와 클릭 이벤트 연결
- 불필요해진 Contract의 메서드 제거

kkosang committed:
feat: 영화 예매 결과에 극장 이름 표시

- 도메인 모델 수정: ReservationMovieInfo에 극장 이름 추가
- TheaterRepository에 극장 이름을 구하는 기능 추가
- MovieDetailPresenterImpl에서 극장 이름을 설정

kkosang committed:
refactor: 영화 예매 결과에 대한 DataBinding

- DataBinding에 필요한 문자열 형식을 전달해주기 위해, UiModel에 출력 문자열 생성 메서드 구현

kkosang committed:
refactor: 불필요한 import 제거

kkosang committed:
docs: STEP 2 기능 요구 사항 작성

kkosang committed:
feat: 하단 네비게이션 바 xml 작성

네비게이션 바를 그리는 데 필요한 리소스와 xml 파일 추가

kkosang committed:
feat: 하단 네비게이션 바 기능 구현

시작 화면 변경
- ScreeningActivity에서 NavigationActivity로 시작 변경

Navigation Activity
- xml 레이아웃 연결
- 각 메뉴에 대한 fragment 연결

Home fragment
- 기존의 ScreeningActivity의 영화 리스트 출력을 Home fragment로 이동

kkosang committed:
refactor: 불필요한 메서드 삭제

kkosang committed:
style: ktLintCheck

kkosang committed:
refactor: fragment xml 수정

Gyuil-Hwnag commented on README.md:
```diff
@@ -1 +1,14 @@
-# android-movie-theater
\ No newline at end of file
+# android-movie-theater
```
README 작성 정말 좋습니다~👍

Gyuil-Hwnag commented on app/build.gradle.kts:
```diff
@@ -2,6 +2,7 @@ plugins {
     id("com.android.application")
     id("org.jetbrains.kotlin.android")
     id("de.mannodermaus.android-junit5") version "1.10.0.0"
+    id("kotlin-kapt")
 }
```
`kotlin-kapt` 플러그인은 추가가 된 이유가 있을까요??🤔

Junyoung-WON replied:
리사이클러 뷰에 대해 데이터바인딩을 하기 위해서 여러 블로그 글을 보다가, 해당 플러그인을 추가하면 경고 메시지가 나타나지 않는다는 것을 보고 별 생각 없이 추가했었습니다... 😞 

지금 확인해보니, 해당 플러그인을 추가하지 않아도 경고 메시지가 나타나지 않네요...!
불필요한 플러그인이니 제거하도록 하겠습니다!

Junyoung-WON replied:
뒤늦게 다시 확인해보니 `kotlin-kapt` 플러그인을 추가하지 않으면 데이터바인딩이 되지 않고, 아래와 같이 컴파일 오류가 나타납니다.
다른 크루들의 build 파일도 살펴보니, 해당 플러그인을 추가한 크루도 있고 추가하지 않은 크루도 있었습니다.
```
error: cannot find symbol
import woowacourse.movie.databinding.ActivityMovieDetailBindingImpl;
                                    ^
  symbol:   class ActivityMovieDetailBindingImpl
  location: package woowacourse.movie.databinding
```
빌드 파일을 확인해보니 해당 DataBinding의 구현체 클래스가 존재하지 않네요.
![image](https://github.com/woowacourse/android-movie-theater/assets/92203597/46a2ce39-e97d-4594-8ab1-6ce5e06deb45)

자세한 이유와 다른 크루들과의 차이는 모르겠지만, `kotlin-kapt` 플러그인을 추가하면 문제없이 잘 동작합니다.
아래 글을 참고하여 추측했을 때는, gradle 빌드 버전이나 설정에서 다른 크루와 차이가 있지 않을까, 아니면 Parcelize 어노테이션 사용에 따라 해당 플러그인이 필요하게 된 것이 아닐까 생각합니다.
[https://3edc.tistory.com/65](https://3edc.tistory.com/65)

Gyuil-Hwnag commented on app/src/androidTest/java/woowacourse/movie/view/MovieDetailActivityTest.kt:
```diff
@@ -0,0 +1,91 @@
+        onView(withId(R.id.summary)).check(matches(withText(SUMMARY)))
+    }
+
+    @Test
+    fun `영화_상세_화면에서_초기_예매_수량은_1이다`() {
+        onView(withId(R.id.reservationInfo))
+            .check(matches(withText("1")))
+    }
```
함수명을 보고 어떤 기능을 테스트 하는지 한눈에 알 수 있군요~👍

Gyuil-Hwnag commented on app/src/androidTest/java/woowacourse/movie/view/MovieDetailActivityTest.kt:
```diff
@@ -0,0 +1,91 @@
+import woowacourse.movie.presentation.view.reservation.detail.MovieDetailActivity
+
+@RunWith(AndroidJUnit4::class)
+class MovieDetailActivityTest {
```
Configuration Change 에 대한 테스트도 있으면 좋을 것 같다는 생각이 듭니다!💪

Gyuil-Hwnag commented on app/src/main/AndroidManifest.xml:
```diff
@@ -13,14 +13,28 @@
                 <category android:name="android.intent.category.LAUNCHER" />
             </intent-filter>
         </activity>
+        <activity
+            android:name=".presentation.view.reservation.seat.SeatSelectionActivity"
+            android:exported="false" />
+        <activity
+            android:name=".presentation.view.reservation.result.ReservationResultActivity"
+            android:exported="false" />
+        <activity
+            android:name=".presentation.view.reservation.detail.MovieDetailActivity"
+            android:exported="false" />
+        <activity
+            android:name=".presentation.view.screening.ScreeningActivity"
+            android:exported="true">
+
+        </activity>
```
패키지 분리 좋습니다~👍

Gyuil-Hwnag commented on app/src/androidTest/java/woowacourse/movie/view/TestFixture.kt:
```diff
@@ -0,0 +1,32 @@
+package woowacourse.movie.presentation.view
```
패키지 이동하면서 해당 파일에 대한 경로 변경 깜빡 하신 것 같습니다!😂

Junyoung-WON replied:
앗... 테스트 코드에 대한 리팩터링을 자주 깜빡하게 되네요ㅠㅠ
movie 패키지의 디렉터리 구조와 동일하게 수정하겠습니다!

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/presentation/uimodel/MovieTicketUiModel.kt:
```diff
@@ -0,0 +1,103 @@
+    val selectedSeats: List<String>,
+    val theaterName: String,
+) : Parcelable {
+    constructor(parcel: Parcel) : this(
+        parcel.readLong(),
+        parcel.readString() ?: "",
+        parcel.readString() ?: "",
+        parcel.readString() ?: "",
+        parcel.readString() ?: "",
+        parcel.readInt(),
+        parcel.readInt(),
+        parcel.readInt(),
+        parcel.createStringArrayList() ?: listOf(),
+        parcel.readString() ?: "",
+    )
+
+    override fun writeToParcel(
+        parcel: Parcel,
+        flags: Int,
+    ) {
+        parcel.writeLong(ticketId)
+        parcel.writeString(title)
+        parcel.writeString(screeningDate)
+        parcel.writeString(startTime)
+        parcel.writeString(endTime)
+        parcel.writeInt(runningTime)
+        parcel.writeInt(reservationCount)
+        parcel.writeInt(totalPrice)
+        parcel.writeStringList(selectedSeats)
+        parcel.writeString(theaterName)
+    }
```
해당 내용은 Parcelable 을 사용하는 모든 객체마다, 보일러 플레이트 코드가 생기겠군요🥲
이를 없앨 수 있는 @Parcelize 라는 어노테이션이 있으므로, 살펴보면 좋을 것 같다는 생각이 듭니다💪

Parcelable 공식문서 : https://developer.android.com/kotlin/parcelize

Junyoung-WON replied:
수업 강의자료에도 언급되어 있었던 내용이었네요...!
자료도 함께 제공해주셔서 감사드립니다!
이 부분도 공부해서 적용해보겠습니다~~ 😄

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/presentation/uimodel/MovieTicketUiModel.kt:
```diff
@@ -0,0 +1,103 @@
+    fun totalPrice(): String = "${String.format("%,d", totalPrice)}원 (현장 결제)"
+
+    companion object CREATOR : Parcelable.Creator<MovieTicketUiModel> {
+        val DEFAULT_DATE_FORMAT: DateTimeFormatter = DateTimeFormatter.ISO_LOCAL_DATE
+        const val DEFAULT_TIME_FORMAT = "HH:mm"
```
해당 내용을 DateUtils 등으로 뽑아서 사용하도록 해보는 것은 어떨까 라는 생각이 듭니다~💪

Junyoung-WON replied:
날짜나 시간 포맷을 변경하는 등 LocalDate 에 관련된 로직이 많아서 코드 작성이 어려웠는데,
관련 기능들을 util 패키지로 분리해서 관리한다면 더 편하겠네요!
util 패키지 활용법을 좀 더 공부해서 리팩터링을 해보겠습니다!

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/presentation/view/bottomNavigationBar/Home.kt:
```diff
@@ -0,0 +1,78 @@
+import woowacourse.movie.presentation.view.screening.ScreeningPresenterImpl
+import woowacourse.movie.presentation.view.screening.adapter.MovieListAdapter
+import woowacourse.movie.presentation.view.screening.theater.TheaterBottomSheetDialogFragment
+
+class Home : Fragment(), ScreeningContract.View, ScreeningContract.ViewActions {
```
Activity 들은 XXActivity 이렇게 하였듯, Fragment도 비슷하게 가보는 것은 어떨까요~💪

Junyoung-WON replied:
Home, Setting 만 보아서는 어떤 역할을 하는 클래스인지 명확하게 알기가 어렵겠네요.
통일성 있게 네이밍을 개선하도록 하겠습니다!

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/presentation/view/bottomNavigationBar/Home.kt:
```diff
@@ -0,0 +1,78 @@
+    }
+
+    override fun showAdContent(content: String) {
+        activity?.runOnUiThread {
+            Toast.makeText(context, content, Toast.LENGTH_SHORT).show()
+        }
```
해당 내용이 들어간 이유가 궁금합니다~🤔

Junyoung-WON replied:
광고를 클릭했을 때의 동작도 간단하게나마 구현해보고 싶었습니다!
영화와 광고 아이템을 나타내고 관리하는 RecyclerView 어댑터에서 각 View의 클릭에 대한 동작을 다르게 설정하고, 클릭에 따른 실질적인 동작은 Activity가 구현하지 않을까 하는 고민에서 이렇게 구현해보았습니다.

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/presentation/view/bottomNavigationBar/NavigationActivity.kt:
```diff
@@ -0,0 +1,53 @@
+        bottomNavigationView.setOnItemSelectedListener { item ->
+            when (item.itemId) {
+                R.id.home_fragment -> {
+                    supportFragmentManager.beginTransaction()
+                        .replace(R.id.selected_fragment, Home())
+                        .addToBackStack(null)
+                        .commit()
+                    true
```
해당 동작은 어떤 동작을 하게 될까요??
~~~kotlin
                    supportFragmentManager.beginTransaction()
                        .replace(R.id.selected_fragment, Home())
                        .addToBackStack(null)
                        .commit()
                    true
~~~
의 구현을 사용하면, Fragment 자체가 교체되는 것이므로, 이전에 대한 상태를 보존하지 못합니다! 와 같이 호두의 자유로운 의견을 공유해주세요💪

추가적으로 `commit()` 과 `commitAllowingStateLoss()` 도 살펴보면 좋을 것 같다는 생각이 듭니다!
(이번 실제 서비스중에서 Crash 로 있었던 이슈라서요!)

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/presentation/view/bottomNavigationBar/ReservationList.kt:
```diff
@@ -0,0 +1,18 @@
+import androidx.fragment.app.Fragment
+import woowacourse.movie.R
+
+class ReservationList : Fragment() {
+    override fun onCreateView(
+        inflater: LayoutInflater,
+        container: ViewGroup?,
+        savedInstanceState: Bundle?,
+    ): View? {
+        return inflater.inflate(R.layout.fragment_reservation_list, container, false)
+    }
+}
```
다음번 요청때는 구현해주실거라 믿겠습니다~😆

Junyoung-WON replied:
넵! 구현해내겠습니다!! 😄

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/presentation/view/bottomNavigationBar/Setting.kt:
```diff
@@ -0,0 +1,18 @@
+import androidx.fragment.app.Fragment
+import woowacourse.movie.R
+
+class Setting : Fragment() {
+    override fun onCreateView(
+        inflater: LayoutInflater,
+        container: ViewGroup?,
+        savedInstanceState: Bundle?,
+    ): View? {
+        return inflater.inflate(R.layout.fragment_setting, container, false)
+    }
+}
```
다음번 요청때는 구현해주실거라 믿겠습니다~😆

Junyoung-WON replied:
설정 화면도 마찬가지로 구현해내겠습니다!

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/presentation/view/reservation/detail/MovieDetailActivity.kt:
```diff
@@ -0,0 +1,177 @@
+import woowacourse.movie.presentation.view.screening.theater.TheaterBottomSheetDialogFragment.Companion.THEATER_ID_KEY
+
+class MovieDetailActivity : BaseActivity(), MovieDetailContract.View {
+    private lateinit var movieDetailPresenter: MovieDetailContract.Presenter
+    private lateinit var timeSpinnerAdapter: ArrayAdapter<String>
+    private lateinit var binding: ActivityMovieDetailBinding
+    private val dateSpinner: Spinner by lazy {
+        findViewById(R.id.dateSpinner)
+    }
+    private val timeSpinner: Spinner by lazy {
+        findViewById(R.id.timeSpinner)
+    }
```
지연초기화를 하는 방법들은 `lateinit var`, `by lazy {}`가 있는데, 각각은 어떤 차이점을 가지고 있으며, 어떤 경우에 사용을 할 때 장단점이 있을까요?🤔

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/presentation/view/reservation/detail/MovieDetailActivity.kt:
```diff
@@ -0,0 +1,177 @@
+            }
+    }
+
+    fun subCount() {
+        movieDetailPresenter.minusReservationCount()
+        updateBinding()
+    }
+
+    fun addCount() {
+        movieDetailPresenter.plusReservationCount()
+        updateBinding()
+    }
```
과연 해당 내용이 Activity 에 있는 것이 맞을까요~🤔

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/presentation/view/reservation/seat/SeatSelectionActivity.kt:
```diff
@@ -0,0 +1,219 @@
+        seatSelectionPresenter.attachView(this)
+
+        confirmButton.isEnabled = false
+        confirmButton.setOnClickListener {
+            showConfirmationDialog()
+        }
```
해당 내용도 BindingAdapter를 사용하도록 해볼 수 있겠다는 생각이 듭니다~💪

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/presentation/view/reservation/seat/SeatSelectionActivity.kt:
```diff
@@ -0,0 +1,219 @@
+        dialog.show()
+    }
+
+    companion object {
+        const val DEFAULT_COUNT = 1
+        const val SEAT_POSITION_TEXT_FORMAT = "%c%d"
+        const val SEAT_ROW_START_VALUE = 'A'
+        const val SEAT_COL_START_VALUE = 1
+        const val FILLED_SEATS_COUNT_MESSAGE = "예매할 좌석 개수를 모두 선택했습니다."
+        const val SHORT_DURATION = 5
+        val B_RANK_COLOR = Color.argb(255, 142, 19, 236)
+        val A_RANK_COLOR = Color.argb(255, 25, 211, 88)
+        val S_RANK_COLOR = Color.argb(255, 27, 72, 233)
+        const val WHITE_COLOR = Color.WHITE
+        const val YELLOW_COLOR = Color.YELLOW
+    }
```
해당 내용들은 android resource 를 활용하도록 해보는 것은 어떨까 라는 생각이 듭니다!💪

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/presentation/view/screening/adapter/MovieListAdapter.kt:
```diff
@@ -0,0 +1,78 @@
+import woowacourse.movie.presentation.view.screening.ScreeningContract
+import woowacourse.movie.presentation.view.screening.adapter.diffutil.MovieDiffCallback
+
+class MovieListAdapter(
+    private var movieList: List<MovieUiModel>,
+    private var adList: List<Ad>,
+    private val listener: ScreeningContract.ViewActions,
+) : RecyclerView.Adapter<ViewHolder>() {
```
해당 Adapter 에서 ad, movie 둘의 리스트를 받는 방법외에 아래와 같이 sealed class로 만드는 방법도 있으니 한번 살펴보는 것도 좋을 것 같다는 생각이 듭니다!💪

- 예시입니다!!
~~~kotlin
sealed class ScreenView {
    data class Ads(val poster: Int) : ScreenView()

    data class Screen(
        val id: Int,
        val movie: Movie,
        val startDate: String,
        val endDate: String,
        val selectableDates: List<ScreenDate>,
    ) : ScreenView()
}
~~~
~~~kotlin
class ScreenRecyclerViewAdapter() : RecyclerView.Adapter<ScreenRecyclerViewAdapter.ScreenRecyclerViewViewHolder>() {

...

override fun getItemViewType(position: Int): Int {
  return when (screens[position]) {
      is Screen -> VIEW_TYPE_SCREEN
      is Ads -> VIEW_TYPE_ADS
      else -> VIEW_TYPE_ADS
  }
}

override fun onCreateViewHolder(): ScreenRecyclerViewViewHolder {
        return when (viewType) {
            VIEW_TYPE_SCREEN -> ScreenRecyclerViewViewHolder.ScreenViewHolder()
            VIEW_TYPE_ADS -> ScreenRecyclerViewViewHolder.AdsViewHolder()

...

sealed class ScreenRecyclerViewViewHolder(view: View): RecyclerView.ViewHolder(view) {
        class AdsViewHolder(val view: View) : ScreenRecyclerViewViewHolder(view) {        }
        class ScreenViewHolder(
            val view: View,
            private val screenActionHandler: ScreenActionHandler,
        ) : ScreenRecyclerViewViewHolder(view) {     }
~~~
예시 코드라서 적용을 하시다가 어려운 부분있으면 언제든지 말씀해주세요!!

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/presentation/view/screening/adapter/MovieListAdapter.kt:
```diff
@@ -0,0 +1,78 @@
+    private val listener: ScreeningContract.ViewActions,
+) : RecyclerView.Adapter<ViewHolder>() {
+    fun updateMovieList(newMovieList: List<MovieUiModel>) {
+        val diffCallback = MovieDiffCallback(movieList, newMovieList)
+        val diffResult = DiffUtil.calculateDiff(diffCallback)
+        movieList = newMovieList
+        diffResult.dispatchUpdatesTo(this)
```
해당 내용처럼 직접 구현을 하는 방법외에 `ListAdapter + DiffUtils` 라는 내용이 있어서, 한번 살펴봐도 좋을 것 같다는 생각이 듭니다~💪

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/presentation/view/screening/adapter/MovieViewHolder.kt:
```diff
@@ -0,0 +1,39 @@
+    fun bind(movie: MovieUiModel) {
+        posterImage.setImageResource(movie.posterImageId)
+        title.text = movie.title
+        screeningDate.text =
+            itemView.context.getString(
+                R.string.screening_date_format,
+                movie.screeningStartDate,
+                movie.screeningEndDate,
+            )
+        runningTime.text =
+            itemView.context.getString(
+                R.string.running_time_format,
+                movie.runningTime,
+            )
```
해당 내용도 BindingAdapter를 사용안다면, ViewHolder 내부의 코드도 줄여볼 수 있겠다는 생각이 듭니다~💪

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/repository/TheaterRepositoryImpl.kt:
```diff
@@ -0,0 +1,212 @@
+import woowacourse.movie.presentation.repository.TheaterRepository
+import woowacourse.movie.presentation.uimodel.TheaterUiModel
+
+object TheaterRepositoryImpl : TheaterRepository {
+    private val dummyTheaters: List<Theater> =
```
미션 제출전에 다음 설정으로 문법에 안 맞는 내용은 없는지 확인해 보시는 것도 좋을것 같습니다~😆
./gradlew clean ktlintCheck 

<img width="1422" alt="스크린샷 2024-05-02 22 49 59" src="https://github.com/woowacourse/android-movie-theater/assets/84956038/804dde5d-ef83-46b9-b6a8-9e0adf592374">

Gyuil-Hwnag commented on app/src/main/res/layout/activity_movie_detail.xml:
```diff
@@ -0,0 +1,192 @@
+
+                <TextView
+                    android:id="@+id/screeningDate"
+                    android:text="@{data.movieUiModel.joinScreeningDates()}"
```
text 에서의 활용 졸습니다~👍

Gyuil-Hwnag commented on app/src/main/res/layout/activity_movie_detail.xml:
```diff
@@ -0,0 +1,192 @@
+                android:layout_height="@dimen/size_45"
+                android:layout_marginStart="@dimen/size_20"
+                android:layout_marginTop="@dimen/size_16"
+                android:onClick="@{()->view.subCount()}"
```
xml 내에서 존재하는 클릭 이벤트에 대한 전체적인 코멘트 입니다!

clickEvent를 받기위해서 현재처럼 Activity 를 받는 방법이 맞는지는 고민해보면 좋을 것 같습니다🥲
순환 의존 구조가 되었다는 생각이 듭니다(Activity <-> xml)

아래와 같이 작성을 하면 재사용및 Activity에 대한 1:1 의존을 안가지도록 작성이 가능해지겠다는 생각이 듭니다!
- 예시입니다!
~~~kotlin
interface ActionHandler {
    fun subCount(
        screenId: Int,
        theaterId: Int,
    )
}

class Presenter() : Contract.Presenter, ActionHandler {
    override fun subCount(screenId: Int, theaterId: Int) {
        TODO("Not yet implemented")
    }
}

<data>
    <variable
        name="actionHandler"
        type="woowacourse.movie.ui.ActionHandler" />

...
android:onClick="@{() -> actionHandler.subCount()}"
~~~

Gyuil-Hwnag reviewed (changes requested):
호두 안녕하세요!
이번 미션을 함께하게 된 두루라고 합니다. 잘부탁드립니다!

해당 미션 구현 고생 많으셨습니다!
미션 구현을 너무 잘 해주셨습니다 💯

DataBinding에 대한 클릭 이벤트에 대한 코멘트는 살펴보시면, 좋을 것 같습니다!
추가로 고민해보시면 좋을만한 내용들 코멘트를 달았으니 확인해주시면 감사하겠습니다.

궁금한 점이나 논의하고 싶은 내용이 있으면 언제든 DM 주세요! 🙏 
화이팅입니다💪

Junyoung-WON committed:
refactor: Parcelize 어노테이션 적용

Junyoung-WON committed:
refactor: androidTest의 패키지 구조를 main 패키지 동일하게 적용

Junyoung-WON committed:
refactor: 모호한 Fragment의 네이밍을 자세하게 수정

Junyoung-WON committed:
refactor: Presenter의 data binding 제거 및 interface의 동작에 대한 binding으로 수정

Contract에 ViewActions interface 추가

xml 태그의 모호한 id 네이밍 수정

spinner에 대한 binding 적용

Junyoung-WON committed:
fix: 예매 개수 텍스트가 변경되지 않는 오류 수정

Junyoung-WON committed:
refactor: 좌석 선택 화면에 data binding 적용

Junyoung-WON committed:
refactor: 영화 목록에서 영화 아이템에 대해 data binding 적용

Junyoung-WON committed:
style: ktlint check 적용

Junyoung-WON commented:
안녕하세요, 두루!!
테코톡 발표준비가 겹쳐서 PR이 늦어졌습니다...! 기다려주셔서 감사드립니다 😭 🙇 
두루가 주신 피드백을 모두 적용하지는 못했고, 우선 data binding을 위주로 개선하고 리팩터링을 진행하였습니다.

리사이클러 뷰에서의 아이템 뷰 타입을 sealed class나 enum class로 관리할 수 있도록 개선하는 것과,
MovieRepository로부터 영화 상영 정보(상영 날짜 및 시간)을 불러오는 기능의 구현을 못했습니다.
그 외에도 부족한 부분이 많은데, 개선이 필요한 부분은 3, 4단계 미션을 진행하며 함께 개선해나가고 싶습니다!

아래는 지난 번 피드백을 받은 코드로부터 지금까지 수정한 내역을 보여주는 커밋 비교 링크입니다.
https://github.com/Junyoung-WON/android-movie-theater/compare/54e964a..f62a78e

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/presentation/uimodel/MovieTicketUiModel.kt:
```diff
@@ -0,0 +1,91 @@
+import woowacourse.movie.presentation.view.reservation.seat.SeatSelectionActivity.Companion.SEAT_ROW_START_VALUE
+import java.time.format.DateTimeFormatter
+
+@Parcelize
```
@Parcelize 추가 좋습니다~👍

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/presentation/uimodel/MovieTicketUiModel.kt:
```diff
@@ -0,0 +1,91 @@
+        val DEFAULT_DATE_FORMAT: DateTimeFormatter = DateTimeFormatter.ISO_LOCAL_DATE
+        const val DEFAULT_TIME_FORMAT = "HH:mm"
+
+        fun from(movieTicket: MovieTicket): MovieTicketUiModel {
+            return MovieTicketUiModel(
+                movieTicket.ticketId,
+                movieTicket.reservationMovieInfo.title,
+                movieTicket.reservationMovieInfo.dateTime.screeningDate.date.format(
+                    DEFAULT_DATE_FORMAT,
+                ),
+                movieTicket.reservationMovieInfo.dateTime.screeningDate.screeningTime.startTime.format(
+                    DateTimeFormatter.ofPattern(DEFAULT_TIME_FORMAT),
+                ),
+                movieTicket.reservationMovieInfo.dateTime.screeningDate.screeningTime.getEndTime()
+                    .format(
+                        DateTimeFormatter.ofPattern(DEFAULT_TIME_FORMAT),
+                    ),
+                movieTicket.reservationMovieInfo.dateTime.screeningInfo.runningTime,
+                movieTicket.reservationInfo.reservationCount,
+                movieTicket.reservationInfo.selectedSeats.totalPrice(),
+                movieTicket.reservationInfo.selectedSeats.seats.map { seat ->
+                    String.format(
+                        SEAT_POSITION_TEXT_FORMAT,
+                        SEAT_ROW_START_VALUE + seat.row,
+                        SEAT_COL_START_VALUE + seat.col,
+                    )
+                },
+                movieTicket.reservationMovieInfo.theaterName,
+            )
```
해당 내용이 들어간 이유가 있는지 궁금합니다!🤔
이미 생성자를 통해서 값을 대입해주고 있는 것 같아서 코멘트 남겨두었습니다~
리팩토링 중이시라, 추후 리팩토링을 위해서 추가된 내용인가 싶기는 하네요😀

Junyoung-WON replied:
나중에 쓰일 일이 있을 것 같아, 혹시나 해서 만들어준 팩토리 함수입니다...!
나중에 리팩터링을 마저 진행하면서 정리하겠습니다!

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/presentation/view/bottomNavigationBar/HomeFragment.kt:
```diff
@@ -0,0 +1,78 @@
+import woowacourse.movie.presentation.view.screening.adapter.MovieListAdapter
+import woowacourse.movie.presentation.view.screening.theater.TheaterBottomSheetDialogFragment
+
+class HomeFragment : Fragment(), ScreeningContract.View, ScreeningContract.ViewActions {
+    private lateinit var adapter: MovieListAdapter
```
네이밍 수정 잘해주셨네요~👍

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/presentation/view/bottomNavigationBar/NavigationActivity.kt:
```diff
@@ -0,0 +1,53 @@
+    private fun setBottomNavigationView() {
+        bottomNavigationView.setOnItemSelectedListener { item ->
+            when (item.itemId) {
+                R.id.home_fragment -> {
+                    supportFragmentManager.beginTransaction()
+                        .replace(R.id.selected_fragment, HomeFragment())
+                        .addToBackStack(null)
+                        .commit()
+                    true
+                }
```
~~~kotlin
                    supportFragmentManager.beginTransaction()
                        .replace(R.id.selected_fragment, HomeFragment())
                        .addToBackStack(null)
                        .commit()
~~~
의 구현을 사용하면, Fragment 자체가 교체되는 것이므로, 이전에 대한 상태를 보존하지 못합니다! 와 같이 호두의 자유로운 의견을 공유해주세요💪

Junyoung-WON replied:
fragment manager의 replace 함수는 기존의 fragment를 제거하고 새로운 fragment를 추가해주는 동작을 합니다!
그래서 이전에 대한 상태를 보존하지 못합니다. 만약 상태를 보존하고자 한다면 DB나 다른 곳에 상태 정보들을 저장하고, Fragment가 새로 교체될 때 저장된 내용들을 불러오는 방법이 있을 것 같습니다.
영화 리스트나 지난 예매 정보가 새로 업데이트되어야 하는 필요성도 있다고 생각해서, 지금의 구현도 일리가 있다고 생각했습니다. 하지만 설정 Fragment의 경우 정적으로 구현되어있는 기능 및 아이템이기 때문에, 매번 새로운 UI를 그리는 것이 불필요한 작업이 될 것 같네요.
조금 더 효율적인 방법이 있는지 찾아보는 것도 좋겠습니다!

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/presentation/view/reservation/detail/MovieDetailActivity.kt:
```diff
@@ -0,0 +1,182 @@
+
+    override fun getLayoutResId(): Int = R.layout.activity_movie_detail
+
+    override fun onCreateSetup(savedInstanceState: Bundle?) {
+        supportActionBar!!.setDisplayHomeAsUpEnabled(true)
```
non-nullable 하다고 !!로 지정을 하는 것은 정말 주의해야 하긴 합니다!!
해당 내용을 좀 더 안전하게 사용을 하려면 어떻게 해야 할지 고민해보면 좋을 것 같다는 생각이 듭니다~💪

Gyuil-Hwnag commented on app/src/main/java/woowacourse/movie/presentation/view/reservation/detail/MovieDetailPresenterImpl.kt:
```diff
@@ -0,0 +1,121 @@
+    val theaterId: Int,
+    private val movieRepository: MovieRepository = MovieRepositoryImpl,
+) : MovieDetailContract.Presenter {
+    private var view: MovieDetailContract.View? = null
+    private val movie: Movie = movieRepository.findMovieById(movieId)
```
view를 꼭 nullable 하다고 설정을 해야할지 고민해보면 좋을 것 같습니다!
nullable하다고 지정을 하여, Presenter 내에서 view를 사용할 때 nullable 한지에 대한 체크를 하는 부분이 들어가고 있는 것 같습니다💪
~~~kotlin
private lateinit var view: MovieDetailContract.View
~~~

Junyoung-WON replied:
lateinit으로 초기화를 늦추는 방법도 있었네요...! 참고해보겠습니다!
그리고 nullable을 무심코 사용하는 경향이 있는데, 내부에서 정확하게 어떤 방식으로 동작하는지 알아볼 필요가 있는 것 같습니다. 어떤 장점이 있고 또 어떤 단점이 존재하는지 찾아보겠습니다!

Gyuil-Hwnag commented on app/src/main/res/layout/activity_movie_detail.xml:
```diff
@@ -0,0 +1,192 @@
+            name="movieDetails"
+            type="woowacourse.movie.presentation.uimodel.MovieUiModel" />
+
+        <variable
+            name="view"
+            type="woowacourse.movie.presentation.view.reservation.detail.MovieDetailContract.ViewActions" />
```
view에 대한 action 지정 좋습니다~
다만, 여러 xml 을 보니 `listener`과 `view`를 섞어서 네이밍이 지정되어 있는 것 같은데, 해당내용을 통합해보는 것도 좋을 것 같다는 생각이 듭니다!💪

Junyoung-WON replied:
detail 페이지에서는 view로, 나머지 다른 곳에서는 listener로 지정했네요..
다른 개발자가 본다면 혼란스러울 수 있으니, 네이밍을 통일시켜보겠습니다!

Gyuil-Hwnag commented on app/src/main/res/layout/activity_movie_detail.xml:
```diff
@@ -0,0 +1,192 @@
+                android:layout_width="@dimen/size_65"
+                android:layout_height="@dimen/size_45"
+                android:layout_marginTop="@dimen/size_16"
+                android:layout_marginEnd="@dimen/size_20"
+                android:onClick="@{()->view.onPlusButtonClicked()}"
```
리뷰 반영 잘해주셨네요~👍

Gyuil-Hwnag commented on app/src/main/res/layout/activity_movie_detail.xml:
```diff
@@ -0,0 +1,192 @@
+                android:layout_marginTop="@dimen/size_16"
+                android:layout_marginEnd="@dimen/size_20"
+                android:onClick="@{()->view.onPlusButtonClicked()}"
+                android:text="@string/plus"
```
android resource의 활용 좋습니다~👍

Gyuil-Hwnag commented on app/src/main/res/values/dimens.xml:
```diff
@@ -0,0 +1,30 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+    <dimen name="size_9">9dp</dimen>
```
해당 처럼 dimen을 사용한다면, 사실상 하드 코딩을 하는 것이랑 별다른 차이점을 못 가질 것 같다는 생각이 듭니다!
아래와 같은 네이밍등으로 좀 더 유의미한 값을 가지도록 해보는 것은 어떨까요?🤔
해당 부분은 무조건적으로 dimen으로 뽑으면 오히려 가독성을 해칠 수 있으니, 호두의 기준을 세우는 것이 좋을 것 같습니다!

(호두가 회사에 갔는데 어떻게 구현되어 있어야, 호두가 좀 더 유지보수가 쉬울 것 같은지를 생각하면서 하면 좀 더 의미있게 와닿을 것 같다는 생각이 듭니다!💪)
~~~kotlin
 <dimen name="notification_linear_layout_height">130dp</dimen>

android:layout_height="@dimen/notification_linear_layout_height"
~~~

Junyoung-WON replied:
dimen의 경우도 유의미한 네이밍을 주어서 관리할 수 있군요!
미션을 구현하는 초기에는 크게 불편함이 없었는데, view가 많아지고 여러 xml을 작성하다보니, 너무 많은 size가 있어 확인하기가 불편했습니다. 시간이 남으면 어떻게 적절하게 dimen 리소스를 정리하고 관리할 수 있을까 고민하면서 리네이밍해보겠습니다!

Gyuil-Hwnag reviewed (approved):
호두 안녕하세요!
해당 미션 너무 고생 많으셨습니다! 
리뷰 반영 정말 잘 해주셨네요💯
DM으로 얘기 나눈대로, 리팩토링을 진행과 3, 4단계를 구현하면 바뀌는 내용이 많을텐데, 화이팅입니다!
(바뀐 내용 링크로 남겨주신 덕분에 빠르게 확인이 가능했습니다👍😆)

해당 미션은 다음 미션 진행을 위하여 머지하도록 하겠습니다.
도움이 될만한 코멘트 달았으니 확인하여 다음 미션에서 반영해주셔도 좋을 것 같습니다.

궁금한 점이나 논의하고 싶은 내용이 있으면 언제든 DM 주세요! 🙏
다음 미션도 화이팅입니다💪

Gyuil-Hwnag commented:
> 안녕하세요, 두루!! 테코톡 발표준비가 겹쳐서 PR이 늦어졌습니다...! 기다려주셔서 감사드립니다 😭 🙇 두루가 주신 피드백을 모두 적용하지는 못했고, 우선 data binding을 위주로 개선하고 리팩터링을 진행하였습니다.
> 
> 리사이클러 뷰에서의 아이템 뷰 타입을 sealed class나 enum class로 관리할 수 있도록 개선하는 것과, MovieRepository로부터 영화 상영 정보(상영 날짜 및 시간)을 불러오는 기능의 구현을 못했습니다. 그 외에도 부족한 부분이 많은데, 개선이 필요한 부분은 3, 4단계 미션을 진행하며 함께 개선해나가고 싶습니다!
> 
> 아래는 지난 번 피드백을 받은 코드로부터 지금까지 수정한 내역을 보여주는 커밋 비교 링크입니다. https://github.com/Junyoung-WON/android-movie-theater/compare/54e964a..f62a78e

커밋 비교 링크 정말 좋습니다ㅎㅎ👍
3, 4단계 구현하면서 리팩토링까지 같이 하시느라 시간적으로 빠듯하다고 느껴지실텐데 화이팅입니다!!💪
